Remove automatic setup feature

jdufresne opened the pull request:
Fixes #838

aaugustin commented:
I'm going to make a release today to fix an incompatibility with a recent release of sqlparse.

I'm not ready to make this change in a hurry, but it's still on the table.

Perhaps you could bring the idea up on django-users?

timgraham commented on debug_toolbar/settings.py:
```diff
@@ -159,13 +158,8 @@ def get_panels():
-
-# The following functions can monkey-patch settings automatically. Several
-# imports are placed inside functions to make it safe to import this module.
+# Several imports are placed inside functions to make it safe to import this
```
It looks like only 1 function is left. Do you understand the comment? It's not clear to me who is doing the importing of this module.

timgraham commented on docs/changes.rst:
```diff
@@ -1,6 +1,19 @@
+Removed features
+~~~~~~~~~~~~~~~~
+
+* Support for automatic setup has been removed. Installation now requires
```
It would be nice to explain why it's been removed.

jdufresne commented:
@matthiask I think  the doc changes from this commit were mistakenly included in 388e177eb19a2b51f7663003a1aa509ee56bbb7a.

matthiask commented:
@jdufresne Oops. Thank you for the report!

matthiask commented:
Request for comments: https://groups.google.com/forum/#!topic/django-users/no_yOMvFZ0M (no replies yet)

Thanks for the feedback @timgraham!

jdufresne commented:
Thanks for the feedback. I believe I have addressed the issues raised. All additional feedback is welcome. I have tested the changes using the project's unit tests and against my own Django applications.

> It looks like only 1 function is left. Do you understand the comment? It's not clear to me who is doing the importing of this module.

In the latest version I have moved `check_middleware()` to `apps.py`. This is the only file that calls this function. Moving it allowed me remove the imports inside the function; they are now at the top level. I removed the comment entirely as I feel it is no longer relevant.

> It would be nice to explain why it's been removed.

I adapted the warnings from the old installation docs. Those previous warnings are now listed as reasons for removing the automatic setup feature in the changes docs. Let me know what you think.

matthiask commented on debug_toolbar/apps.py:
```diff
@@ -1,16 +1,46 @@
+    debug_toolbar_index = None
+
+    # Determine the indexes which gzip and/or the toolbar are installed at
+    for i, middleware in enumerate(settings.MIDDLEWARE_CLASSES):
```
MIDDLEWARE should also be checked, not only MIDDLEWARE_CLASSES

jdufresne replied:
> IMO the check could also be replaced by a big warning in the installation docs, but not sure whether that would be alright with everyone.

This already exists to some extent. From the [installation docs Middleware section](https://raw.githubusercontent.com/jazzband/django-debug-toolbar/master/docs/installation.rst):

> The order of ``MIDDLEWARE_CLASSES`` is important. You should include the Debug
Toolbar middleware as early as possible in the list. However, it must come
after any other middleware that encodes the response's content, such as
:class:`~django.middleware.gzip.GZipMiddleware`.

Is this sufficient? Or would you like me to expand on this blurb? Or should I continue with modifying the runtime check?

matthiask commented:
I feel a bit bad making those judgement calls seeing as I'm not contributing to DDT for a long time (at all). But let's continue this :-)

I very much prefer checking MIDDLEWARE first, since I don't think that removing the check removes much complexity from the codebase, and it's one stumbling block less when adding DDT to a project.

jdufresne commented:
* Updated the runtime check to handle `MIDDLEWARE`
* Updated the runtime check to warn on the absence of `DebugToolbarMiddleware` (as it is no longer added automatically)
* Updated the warning wording based on removal of automatic setup
* Emphasized the `GZipMiddleware` documentation as a  "warning"

matthiask commented:
Thank you! Now everything looks good to me. I'm not worried about the unsuccessful codecov/patch check in this case.

Let's wait a working day or two for feedback on the mailing list and here on the pull request. I'll merge in case nobody brings up some good points why this shouldn't be done.

Thanks again!

jdufresne commented:
I only just now realized that DDT includes an [example project](https://github.com/jazzband/django-debug-toolbar/tree/master/example) (for testing?). Would you like this updated as well to use the explicit setup? Should be easy to update.

aaugustin commented:
Yes, the example project is for manual testing and should be updated to reflect whatever the documentation recommends.

jdufresne commented:
Thanks. The example project has been updated to current Django and DDT configuration recommendations.

timgraham commented on debug_toolbar/apps.py:
```diff
@@ -1,16 +1,58 @@
+        check_middleware()
+
+
+def check_middleware():
```
If not too complicated, that might be a candidate for using the system check framework.

timgraham commented on docs/changes.rst:
```diff
@@ -4,6 +4,27 @@ Change log
+Removed features
+~~~~~~~~~~~~~~~~
+
+* Support for automatic setup has been removed. The automatic setup was
```
I only expected something like "as it was frequently problematic...." without all the detailed bullet points. It might be useful to include these details on the issue in case someone complains and wants to see the detailed justification, but it's not so important to mention in the release notes, I think.

Although, I have a small/medium django-cms project that seems to use the automatic setup (not sure I'd describe as a "trivial project". ;-)

timgraham commented on example/settings.py:
```diff
@@ -11,26 +11,33 @@
 
 DEBUG = True
 
+INTERNAL_IPS = [
```
Maybe a single line?

jdufresne commented:
> If not too complicated, that might be a candidate for using the system check framework.

Great idea! I have updated the change the use system checks. I also added a test for these systems checks so codecov is happy again.

I made other suggested changes as well. Thanks!
